### PR TITLE
🌱 ClusterClass: Add topology field for NodeDrainTimeout

### LINKED ISSUE #6278: ClusterClass: make MachineDeployment.spec.template.spec.nodeDrainTimeOut patchable
**User Story**

As a [developer/user/operator] I would like to patch `MachineDeployment` for configuring `nodeDrainTimeOut` for workers.

Right now, patching `MachineDeploymentClass` allows us to patch `InfraMachineTemplate` and `KubeadmConifgTemplate`. But `nodeDrainTimeOut` is defined in `MachineTemplate` of `MachineDeployment`, so requesting patching it.

/kind feature

### PR DESCRIPTION
**Pull request by weikaipan**
**What this PR does / why we need it**:

Make MachineDeployment.spec.template.spec.nodeDrainTimeOut patchable in ClusterClass

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6278

### CODE REVIEW

File: internal/controllers/topology/cluster/desired_state.go [hunk 1]
```diff
@@ -212,6 +212,13 @@ func computeControlPlane(_ context.Context, s *scope.Scope, infrastructureMachin
 		}
 	}
 
+	// If it is required to manage the number of replicas for the control plane, set the corresponding field.
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/desired_state.go:215]:
```suggestion
	// If it is required to manage the NodeDrainTimeout for the control plane, set the corresponding field.
```
nit

File: internal/controllers/topology/cluster/desired_state.go [hunk 2]
```diff
@@ -212,6 +212,13 @@ func computeControlPlane(_ context.Context, s *scope.Scope, infrastructureMachin
...
+	if s.Blueprint.Topology.ControlPlane.NodeDrainTimeout != nil {
+		if err := contract.ControlPlane().NodeDrainTimeout().Set(controlPlane, *s.Blueprint.Topology.ControlPlane.NodeDrainTimeout); err != nil {
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/desired_state.go:217]:
```suggestion
		if err := contract.ControlPlane().MachineTemplate().NodeDrainTimeout().Set(controlPlane, *s.Blueprint.Topology.ControlPlane.NodeDrainTimeout); err != nil {
```
see: https://cluster-api.sigs.k8s.io/developer/architecture/controllers/control-plane.html#required-spec-fields-for-implementations-using-machines

File: internal/contract/controlplane.go [hunk 3]
```diff
@@ -76,6 +76,13 @@ func (c *ControlPlaneContract) Replicas() *Int64 {
 	}
 }
 
+// NodeDrainTimeout provides access to the nodeDrainTimeout of a ControlPlane.
+func (c *ControlPlaneContract) NodeDrainTimeout() *Duration {
```
Comment (sbueringer) [inline internal/contract/controlplane.go:80]: Can be removed as we now use `contract.ControlPlane().MachineTemplate().NodeDrainTimeout()` instead

File: internal/controllers/topology/cluster/patches/engine.go [hunk 4]
```diff
@@ -327,6 +327,7 @@ func updateDesiredState(ctx context.Context, req *api.GenerateRequest, blueprint
 		contract.ControlPlane().MachineTemplate().Metadata().Path(),
 		contract.ControlPlane().MachineTemplate().InfrastructureRef().Path(),
 		contract.ControlPlane().Replicas().Path(),
+		contract.ControlPlane().NodeDrainTimeout().Path(),
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/patches/engine.go:330]:
```suggestion
		contract.ControlPlane().MachineTemplate().NodeDrainTimeout().Path(),
		contract.ControlPlane().Replicas().Path(),
```

File: internal/controllers/topology/cluster/desired_state_test.go [hunk 5]
```diff
@@ -282,6 +285,7 @@ func TestComputeControlPlane(t *testing.T) {
 
 		assertNestedField(g, obj, version, contract.ControlPlane().Version().Path()...)
 		assertNestedField(g, obj, int64(replicas), contract.ControlPlane().Replicas().Path()...)
+		assertNestedField(g, obj, string(duration), contract.ControlPlane().NodeDrainTimeout().Path()...)
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/desired_state_test.go:288]:
```suggestion
		assertNestedField(g, obj, fmt.Sprintf("%q", duration), contract.ControlPlane().MachineTemplate().NodeDrainTimeout().Path()...)
```

**Review by sbueringer — COMMENTED** [review]
Looks pretty good already. Please re-run conversion-gen (see diff in the verify job).
I think that this might be necessary because after the Convert func is added conversion-gen adapts to that.

The test failure is due to:
```
WARN invalid TestEvent: FAIL	sigs.k8s.io/cluster-api/internal/controllers/topology/cluster [build failed]
bad output from test2json: FAIL	sigs.k8s.io/cluster-api/internal/controllers/topology/cluster [build failed]

# sigs.k8s.io/cluster-api/internal/controllers/topology/cluster
internal/controllers/topology/cluster/desired_state_test.go:288:29: conversion from Duration (int64) to string yields a string of one rune, not a string of digits (did you mean fmt.Sprint(x)?)
```
>  https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/6379/pull-cluster-api-test-main/1511612926618767360/build-log.txt

**Comment by weikaipan** [comment]
> Looks pretty good already. Please re-run conversion-gen (see diff in the verify job). I think that this might be necessary because after the Convert func is added conversion-gen adapts to that.
> 
> The test failure is due to:
> 
> ```
> WARN invalid TestEvent: FAIL	sigs.k8s.io/cluster-api/internal/controllers/topology/cluster [build failed]
> bad output from test2json: FAIL	sigs.k8s.io/cluster-api/internal/controllers/topology/cluster [build failed]
> 
> # sigs.k8s.io/cluster-api/internal/controllers/topology/cluster
> internal/controllers/topology/cluster/desired_state_test.go:288:29: conversion from Duration (int64) to string yields a string of one rune, not a string of digits (did you mean fmt.Sprint(x)?)
> ```
> 
> > https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/6379/pull-cluster-api-test-main/1511612926618767360/build-log.txt

I ran conversion-gen but it didn't make changes to code.

File: internal/controllers/topology/cluster/desired_state.go [hunk 6]
```diff
@@ -212,6 +212,13 @@ func computeControlPlane(_ context.Context, s *scope.Scope, infrastructureMachin
 		}
 	}
 
+	// If it is required to manage the NodeDrainTimeout for the control plane, set the corresponding field.
+	if s.Blueprint.Topology.ControlPlane.NodeDrainTimeout != nil {
+		if err := contract.ControlPlane().MachineTemplate().NodeDrainTimeout().Set(controlPlane, *s.Blueprint.Topology.ControlPlane.NodeDrainTimeout); err != nil {
+			return nil, errors.Wrap(err, "failed to set spec.nodeDrainTimeout in the ControlPlane object")
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/desired_state.go:218]:
```suggestion
			return nil, errors.Wrap(err, "failed to set spec.machineTemplate.nodeDrainTimeout in the ControlPlane object")
```

**Review by sbueringer — COMMENTED** [review]
last nit + we have to make the verify job work. Apart from that lgtm from my side

**Comment by sbueringer** [comment]
@weikaipan I really have no idea what is going on with conversion-gen. On my machine I end up with this delta: https://github.com/sbueringer/cluster-api/commit/d206569085ef4a6e91847273f6c78ce182855edb

**Comment by sbueringer** [comment]
/retitle 🌱 ClusterClass: Add topology field for NodeDrainTimeout
(I think it's fine to un-WIP, we're close to ready for merge)

**Comment by sbueringer** [comment]
@weikaipan Yup looks perfect. Thx!

Last comment from my side. Can you please squash the commits?

We are running Prow in merge-mode so our release-note tool works, but this has the effect that all commits of a PR are preserved when merging (=> we have to squash them manually before merge)

**Review by killianmuldoon — APPROVED** [review]
/lgtm

**Review by fabriziopandini — COMMENTED** [review]
Congrats for your first PR in CAPI @weikpan, great work!
/lgtm
/approve

**Pull request merged**